Connections: New "Connect data" page with data source catalog

mikkancso opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

We're building out a new Section called Connections behind the feature toggle `dataConnectionsConsole`.
In this section we will have this "Connect data" page, where the available data sources are listed.
We'd like to build a very similar page to this in the cloud-onboarding plugin, so I tried to create
separate, potentially reusable components:
- Search
- CategoryHeader
- CardGrid
- NoResults

This way the ConnectData page's responsibility is only to fetch data and to combine the above components in a meaningful way.

We still have to discuss whether we'd want to make these components available in `@grafana/ui` or not. 

![Screenshot from 2022-10-18 13-48-47](https://user-images.githubusercontent.com/13637610/196421475-3e5ad0e0-7d63-4e97-bbd3-50f20aa63da4.png)


**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Relates to #2273

**Special notes for your reviewer**:

mikkancso committed:
remove Plugins and CloudIntegrations tab and add ConnectData tab

mikkancso committed:
ConnectData: add Search component and use it

mikkancso committed:
ConnectData: add DataSourcePluginList component

mikkancso committed:
add CardGrid component

mikkancso committed:
add CategoryHeader component

mikkancso committed:
ConnectData: restructure content

DataSourcePluginList is removed, because its responsibilities are
actually the same as ConnectData's responsibilities.

NoResults was added as a reusable component, and was moved out of
CardGrid, since there could be more CardGrid on one page, but only one
NoResults.

mikkancso committed:
fix spacer

mikkancso committed:
use LoadingPlaceholder

mikkancso committed:
CardGrid: add margin

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=56772) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=56772) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=56772) it.</sub>

mikkancso committed:
generalize CardGridProps

mikkancso committed:
move isLoading and error into CardGrid

We'd like CardGrid to be reusable, even multiple times within a page.
In this case, it's better UX if we show the loading or error states per
card grid, not for the whole page.

mikkancso committed:
ConnectData: fix NoResults condition

mikkancso committed:
fix and add meaningful tests

mikkancso committed:
fix indentation

leventebalogh commented on public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx:
```diff
@@ -0,0 +1,74 @@
+  sourcesList: css`
+    display: grid;
+    grid-template-columns: repeat(auto-fill, minmax(280px, 1fr));
+    gap: 12px;
```
Nit: maybe we should the `theme.spacing(1.5)` here? Although as we would need to use "1.5" maybe it doesn't make sense.

leventebalogh commented on public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx:
```diff
@@ -0,0 +1,74 @@
+
+export interface CardGridProps {
+  items: Array<{ id: string; name: string; logo?: string }>;
+  isLoading: boolean;
```
Would it make sense to make the `isLoading` prop optional so in case it's not passed in then the default state is just "not loading"?

leventebalogh commented on public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx:
```diff
@@ -0,0 +1,74 @@
+  return isLoading ? (
+    <LoadingPlaceholder text="Loading..." />
+  ) : !!error ? (
+    <div>Error: {error.message}</div>
```
As we are not using anything else (e.g. stack trace) from the error object apart from the `.message` field, it might makes sense to just have a `errorMessage?: string` prop instead of `error: any`. This could make the logic and the typing a bit simpler.

Thinking about it a bit more, maybe it's more the responsibility of `<ConnectData />` to handle any loading errors, as the async fetching actually happens there (and that is also the place where we are handling the case of no results). 

What do you think?

mikkancso replied:
The original reason why I put `error` in the `CardGrid` was:
> We'd like CardGrid to be reusable, even multiple times within a page.
In this case, it's better UX if we show the loading or error states per
card grid, not for the whole page.

But yeah, it actually feels like it doesn't belong there. And pages _could_ do something like: 
```
isCardGrid1Loading ? (
    <LoadingPlaceholder text="Loading..." />
  ) : !!errorCardGrid1 ? (
    <div>Error: {errorCardGrid1.message}</div>
) : <CardGrid one>

isCardGrid2Loading ? (
    <LoadingPlaceholder text="Loading..." />
  ) : !!errorCardGrid2 ? (
    <div>Error: {errorCardGrid2.message}</div>
) : <CardGrid two>
```
It feels like a lot of boilerplate, but actually it's better separation of concerns. :+1:

leventebalogh commented on public/app/features/connections/tabs/ConnectData/CardGrid/CardGrid.tsx:
```diff
@@ -0,0 +1,74 @@
+  ) : (
+    <ul className={styles.sourcesList}>
+      {items.map((item) => (
+        <Card key={item.id} className={styles.card} href={`plugins/${item.id}`}>
```
Should we make the URL (href) come from the item object as well in case we would like to make this component more reusable?

leventebalogh reviewed (approved):
Left a few notes, but generally LGTM 👍 

(I think we will probably need to rename / prefix the building block components when we would like to make them available publicly, but that can be done later.)

mikkancso committed:
move isLoading and error back to ConnectData

mikkancso committed:
make `url` required for CardGrid items

leventebalogh reviewed (approved):
🚀

torkelo reviewed (commented):
test this with topnav feature toggle enabled. 

The Page component, you should not pass a custom navModel to it for base navtree pages but instead just pass a navId (baseId -connect-data)

for item pages (integrations/data sources) you can pass in the pageNav property to the Page component

leventebalogh commented:
Thanks @torkelo, yeah, that's a good point, we are planning to address that in [this upcoming PR](https://github.com/grafana/grafana/pull/56373), hope that's fine with you.
